### DISCUSSION TITLE
Any way to de/serialize SqlAlchemy 2.0 metadata?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [sqlalchemy/sqlalchemy#11360] [Usage Questions]
I know this has come up a couple times and it seems like the answer is this isn't possible any more, but I thought I'd check anyway.

My use case is we have a repo that represents our database via SqlAlchemy models. We use alembic to create the database and apply changes. In our CI pipeline we run the alembic command many times to do lots of checks that new revisions haven't introduced subtle errors. Each time the alembic command runs it must rebuild the SqlAlchemy metadata and this end up taking up a significant amount of the CI time (nearly 50%). 

In SqlAlchemy 1.4 I am able to use cloudpickle to serialize the metadata and speed up our CI run ~40%, but in SqlAlchemy 2.0 I get an error like this:

```
Traceback (most recent call last):
  File ".../bin/alembic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File ".../python3.11/site-packages/alembic/config.py", line 641, in main
    CommandLine(prog=prog).main(argv=argv)
  File ".../python3.11/site-packages/alembic/config.py", line 631, in main
    self.run_cmd(cfg, options)
  File ".../python3.11/site-packages/alembic/config.py", line 608, in run_cmd
    fn(
  File ".../python3.11/site-packages/alembic/command.py", line 290, in check
    script_directory.run_env()
  File ".../python3.11/site-packages/alembic/script/base.py", line 583, in run_env
    util.load_python_file(self.dir, "env.py")
  File ".../python3.11/site-packages/alembic/util/pyfiles.py", line 95, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../python3.11/site-packages/alembic/util/pyfiles.py", line 113, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "./migrations/env.py", line 366, in <module>
    run_migrations_online()
  File "./migrations/env.py", line 350, in run_migrations_online
    target_metadata=get_target_metadata(engine),
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./migrations/env.py", line 81, in get_target_metadata
    cloudpickle.dump(strike_models.base.metadata, f)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../python3.11/site-packages/cloudpickle/cloudpickle.py", line 1461, in dump
    Pickler(file, protocol=protocol, buffer_callback=buffer_callback).dump(obj)
  File ".../python3.11/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
           ^^^^^^^^^^^^^^^^^
  File ".../python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 1308, in __getattr__
    value = getattr(self, f"_memoized_attr_{key}")()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../python3.11/site-packages/sqlalchemy/event/attr.py", line 81, in _memoized_attr_ref
    return weakref.ref(self, registry._collection_gced)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: cannot create weak reference to '_JoinedListener' object
```

It seems SqlAlchemy's usage of weakrefs breaks the assumptions of cloudpickle. Is there any obvious workaround?

**Comment by zzzeek** [comment]
Table metadata should be serializable via pickle.  I dont know what cloudpickle is.  if you can reproduce your issue using Python pickle please illustrate a model that has this problem.

**Reply by zzzeek** [reply]
the _JoinedListener code shown there is not used in a schema structure.  it's used by a `Connection` with an `Engine`, which should not be part of a schema structure (make sure you don't have code like `metadata.bind = engine` somewhere), and it's used by the `Enum` and `Boolean` datatypes when they are dialect-adapted, however dialect-adapted types are not stored with the table metadata structure.

**Reply by notatallshaw-gts** [reply]
The main difference between cloudpickle and pickle is that cloudpickle can serialize a function or class by value, whereas pickle can only serialize it by reference.

Some of the sqlalchemy metadata I generate is dynamically constructed, and I get an error when dynamically defining an enum column type and then trying to use pickle (or dill) to serialize the metadata, hence why cloudpickle was my workaround for SqlAlchemy 1.4.

Here is a minimally reproducible example of failing to use regular pickle:

```python
import enum
import pickle

from sqlalchemy import Column, Integer, PrimaryKeyConstraint, Table
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.types import Enum

# Get base and metadata
base = declarative_base()
sa_metadata = base.metadata


# Construct table with enum
id_column = Column("id", Integer())
primary_key = PrimaryKeyConstraint("id", name="id_PK")
field1_column = Column(
    "field1",
    type_=Enum(enum.Enum("field1", ["red", "blue", "green"])),
)
Table("foo", sa_metadata, id_column, primary_key, field1_column, schema="foo")

# Pickle metadata
with open("test.pkl", "wb") as f:
    pickle.dump(sa_metadata, f)
```

With the error:

```
Traceback (most recent call last):
  File "example.py", line 33, in <module>
    pickle.dump(sa_metadata, f)
_pickle.PicklingError: Can't pickle <enum 'field1'>: attribute lookup field1 on __main__ failed
```

**Reply by zzzeek** [reply]
cloudpickle seems to use a real pickler so I still need to see how you are getting a _JoinedListener to be present within a MetaData structure. just as a note this object has to do with event.listen, which can be used explicitly, and it's also used internally by types such as Enum and Boolean, however the _JoinedListener part should not come into play and I cannot get a reproduction case here.

**Reply by notatallshaw-gts** [reply]
> and it's used by the `Enum` and `Boolean` datatypes when they are dialect-adapted, however dialect-adapted types are not stored with the table metadata structure.

Aha, I think this might be the issue! We are serializing dialect specific metadata, specifically where it is required to accurately represent MySQL integer columns, e.g. signed vs. unsigned, and tiny, small, medium, and big ints.

But it looks like in a few places MySQL specific `ENUM`s have leaked in where they shouldn't have. I will attempt cleaning up all of these and see if it makes a difference.

**Reply by notatallshaw-gts** [reply]
Removing all cases of `mysql.ENUM` unfortunately did not help. I will work on producing a minimally reproducible example but it might take a day or two.

Thanks a lot for your analysis, I will perhaps solve my issue by attempting to create a reproducible example, either way I will post my findings back here.

**Reply by zzzeek** [reply]
the mysql.ENUM should not have any problems either

**Reply by notatallshaw-gts** [reply]
This is the smallest reproducible example I was able to make, specifically when you subclass an abstract model that contains an enum column:

```python
import cloudpickle
import sqlalchemy as sa
from sqlalchemy.orm.decl_api import declarative_base

# Get base and metadata
base = declarative_base()
sa_metadata = base.metadata

class _Foo(base):
    __abstract__ = True
    __tablename__ = "foo"

    id_ = sa.Column(sa.Integer(), primary_key=True)
    column1 = sa.Column(sa.Enum("red", "green", name="column1"))

class Foo(_Foo):
    pass

# Pickle metadata
with open("test.pkl", "wb") as f:
    cloudpickle.dump(sa_metadata, f)
```

I've experimented with a few different approaches to adding the enum column and I get the same exception each time.

**Reply by notatallshaw-gts** [reply]
It is also not regular pickleable, but with a different exception:

```
Traceback (most recent call last):
  File "/example.py", line 25, in <module>
    pickle.dump(sa_metadata, f)
_pickle.PicklingError: Can't pickle <class 'sqlalchemy.event.base.JoinedDDLEventsDispatch'>: attribute lookup JoinedDDLEventsDispatch on sqlalchemy.event.base failed
```

Again, only when the enum is in the abstract class, if it is in the concrete class it works fine. It seems whatever magic is allowing enums to be pickled as class attributes is failing when they are subclassed.

**Reply by zzzeek** [reply]
OK there might be some event things happneing, not sure

**Reply by zzzeek** [reply]
oh I know why, because the Enum on the base is in fact doing an adaption.  so that's a bug, that's all, thanks for the repro

**Reply by zzzeek** [reply]
#11365

**Reply by notatallshaw-gts** [reply]
FYI, I've created a follow up issue, `pickle.dump` works now, but I am getting an error with `pickle.load` in a fresh process: https://github.com/sqlalchemy/sqlalchemy/issues/11530
